File: app/src/main/java/com/solution/it/newsoft/ui/ListingActivity.java
package com.solution.it.newsoft.ui;

import android.app.Dialog;
import android.app.ProgressDialog;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Toast;

import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import dagger.android.support.DaggerAppCompatActivity;

import com.google.android.material.snackbar.Snackbar;
import com.solution.it.newsoft.R;
import com.solution.it.newsoft.databinding.ActivityListingBinding;
import com.solution.it.newsoft.databinding.DialogUpdateListBinding;
import com.solution.it.newsoft.model.List;
import com.solution.it.newsoft.paging.ListingAdapter;
import com.solution.it.newsoft.viewmodel.ListingViewModel;

import javax.inject.Inject;

public class ListingActivity extends DaggerAppCompatActivity {
    private ActivityListingBinding binding;
    private ListingViewModel listingViewModel;
    private Toast toast;
//    private Snackbar snackbar;

    @Inject
    ViewModelProvider.Factory viewModelFactory;

    @Inject
    ListingAdapter adapter;

    @Inject
    SharedPreferences prefs;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_listing);

        getSupportActionBar().setDisplayHomeAsUpEnabled(false);
        getSupportActionBar().setDisplayShowHomeEnabled(false);
        getSupportActionBar().setTitle(prefs.getString(ListingViewModel.USERNAME, ""));

        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        binding.recyclerView.setLayoutManager(layoutManager);
        binding.recyclerView.setHasFixedSize(true);

        binding.recyclerView.setAdapter(adapter);

        listingViewModel = ViewModelProviders.of(this, viewModelFactory).get(ListingViewModel.class);
        binding.swipeRefresh.setRefreshing(true);
        listingViewModel.getListLiveData().observe(this, lists -> adapter.submitList(lists));
        listingViewModel.getNetworkState().observe(this, networkState -> {
            adapter.setNetworkState(networkState);
            if (binding.swipeRefresh.isRefreshing())
                binding.swipeRefresh.setRefreshing(false);
        });

        binding.swipeRefresh.setColorSchemeColors(getColor(R.color.colorPrimary));
        binding.swipeRefresh.setOnRefreshListener(() -> refreshList());

        adapter.setOnItemClickListener(new ListingAdapter.OnItemClickListener() {
            @Override
            public void onItemClick(List list) {
//                snackbar = Snackbar.make(binding.coordinatorLayout,
//                        new StringBuilder("List name: ").append(list.getList_name())
//                        .append("\n")
//                        .append("Distance: ").append(list.getDistance()), Snackbar.LENGTH_SHORT);
//                snackbar.show();

                if (toast != null) toast.cancel();
                toast = Toast.makeText(ListingActivity.this,
                        new StringBuilder("List name: ").append(list.getList_name())
                                .append("\n")
                                .append("Distance: ").append(list.getDistance()), Toast.LENGTH_SHORT);
                toast.show();
            }

            @Override
            public void onItemLongClick(List list, int position) {
                showUpdateDialog(list, position);
            }

            @Override
            public void onRetryClick() {
                listingViewModel.retry();
            }
        });
    }

    private void refreshList() {
        binding.swipeRefresh.setRefreshing(true);
        listingViewModel.reload();
    }

    private void showUpdateDialog(List list, int position) {
        Dialog dialog = new Dialog(ListingActivity.this);
        DialogUpdateListBinding dialogBinding = DialogUpdateListBinding.inflate(LayoutInflater.from(ListingActivity.this), (ViewGroup) binding.getRoot(), false);
        dialogBinding.setList(list);
        dialogBinding.etListName.requestFocus();
        dialogBinding.btnUpdate.setOnClickListener(view -> updateList(dialogBinding, list, position, dialog));

        dialog.setContentView(dialogBinding.getRoot());
        dialog.show();

        Window window = dialog.getWindow();
        window.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE);
    }

    private void updateList(DialogUpdateListBinding dialogBinding, List list, int position, Dialog dialog) {
        ProgressDialog progress = ProgressDialog.show(ListingActivity.this, "", "Updating...", true);
        progress.show();
        LiveData<Boolean> isUpdated = listingViewModel.updateList(list.getId(), dialogBinding.etListName.getText().toString(),
                dialogBinding.etDistance.getText().toString());
        isUpdated.observe(ListingActivity.this, aBoolean -> {
            if (aBoolean) {
                adapter.updateList(position, dialogBinding.etListName.getText().toString(),
                        dialogBinding.etDistance.getText().toString());
            }
            progress.dismiss();
            dialog.dismiss();
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater menuInflater = getMenuInflater();
        menuInflater.inflate(R.menu.main_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.logout:
                prefs.edit().clear().apply();
                Intent intent = new Intent(ListingActivity.this, LoginActivity.class);
                startActivity(intent);
                Toast.makeText(this, "Logged out", Toast.LENGTH_SHORT).show();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        finishAffinity();
    }
}

File: app/src/main/java/com/solution/it/newsoft/ui/ListingActivity.kt
package com.solution.it.newsoft.ui

import android.app.Dialog
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity

import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.lifecycleScope
import androidx.paging.LoadState
import androidx.recyclerview.widget.LinearLayoutManager
import com.github.razir.progressbutton.attachTextChangeAnimator
import com.github.razir.progressbutton.bindProgressButton
import com.github.razir.progressbutton.hideProgress
import com.github.razir.progressbutton.showProgress
import io.reactivex.Completable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers

import com.google.android.material.snackbar.Snackbar
import com.solution.it.newsoft.R
import com.solution.it.newsoft.databinding.ActivityListingBinding
import com.solution.it.newsoft.databinding.DialogUpdateListBinding
import com.solution.it.newsoft.model.List
import com.solution.it.newsoft.model.NetworkState
import com.solution.it.newsoft.paging.FooterAdapter
import com.solution.it.newsoft.paging.ListingAdapter
import com.solution.it.newsoft.viewmodel.ListingViewModel
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import java.util.*

import java.util.concurrent.TimeUnit
import javax.inject.Inject
import kotlin.concurrent.schedule

@ExperimentalCoroutinesApi
@AndroidEntryPoint
class ListingActivity : AppCompatActivity() {
    private lateinit var binding: ActivityListingBinding
    private lateinit var layoutManager: LinearLayoutManager
    private val disposable = CompositeDisposable()
    private lateinit var snackbar: Snackbar

    private val viewModel: ListingViewModel by viewModels()
    @Inject lateinit var adapter: ListingAdapter
    @Inject lateinit var prefs: SharedPreferences

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = DataBindingUtil.setContentView(this, R.layout.activity_listing)

        supportActionBar?.setDisplayHomeAsUpEnabled(false)
        supportActionBar?.setDisplayShowHomeEnabled(false)
        supportActionBar?.title = prefs.getString(ListingViewModel.USERNAME, "")

        layoutManager = LinearLayoutManager(this)
        binding.recyclerView.layoutManager = layoutManager
        binding.recyclerView.setHasFixedSize(true)

        val footerAdapter = FooterAdapter(adapter)
        binding.recyclerView.adapter = adapter.withLoadStateFooter(footerAdapter)

        binding.swipeRefresh.isRefreshing = true

        lifecycleScope.launch {
            viewModel.flow.collectLatest { pagingData ->
                binding.swipeRefresh.isRefreshing = false
                adapter.submitData(pagingData)
            }
        }

        binding.swipeRefresh.setColorSchemeColors(getColor(R.color.colorPrimary))
        binding.swipeRefresh.setOnRefreshListener { refreshList() }

        adapter.setOnItemClickListener(object : ListingAdapter.OnItemClickListener {
            override fun onItemClick(list: List) {
                snackbar = Snackbar.make(binding.coordinatorLayout,
                        StringBuilder("List name: ").append(list.list_name)
                                .append("\n")
                                .append("Distance: ").append(list.distance), Snackbar.LENGTH_SHORT)
                snackbar.show()
            }

            override fun onItemLongClick(list: List, position: Int) {
                showUpdateDialog(list, position)
            }
        })
    }

    private fun refreshList() {
        adapter.refresh()
    }

    private fun showUpdateDialog(list: List, position: Int) {
        val dialog = Dialog(this)
        val dialogBinding = DialogUpdateListBinding.inflate(LayoutInflater.from(this), binding.root as ViewGroup, false)
        dialogBinding.list = list
        dialogBinding.etListName.requestFocus()
        dialogBinding.btnUpdate.setOnClickListener { updateList(dialogBinding, list, position, dialog) }

        dialog.setContentView(dialogBinding.root)
        dialog.show()

        val window = dialog.window
        window?.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        window?.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE)
    }

    private fun updateList(dialogBinding: DialogUpdateListBinding, list: List, position: Int, dialog: Dialog) {
        bindProgressButton(dialogBinding.btnUpdate)
        dialogBinding.btnUpdate.attachTextChangeAnimator()
        dialogBinding.btnUpdate.showProgress {
            buttonText = "Updating"
            progressColorRes = R.color.colorPrimary
        }

        val isUpdated = viewModel.updateList(list.id, dialogBinding.etListName.text.toString(),
                dialogBinding.etDistance.text.toString())
        isUpdated.observe(this, Observer { aBoolean ->
            if (aBoolean) {
                adapter.updateList(position, dialogBinding.etListName.text.toString(),
                        dialogBinding.etDistance.text.toString())
            }
            dialogBinding.btnUpdate.hideProgress("Updated")
            Timer().schedule(1_000) {
                dialog.dismiss()
            }
        })
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val menuInflater = menuInflater
        menuInflater.inflate(R.menu.main_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.logout -> {
                prefs.edit().clear().apply()
                val intent = Intent(this, LoginActivity::class.java)
                startActivity(intent)
                Toast.makeText(this, "Logged out", Toast.LENGTH_SHORT).show()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onBackPressed() {
        if (layoutManager.findFirstVisibleItemPosition() > 5) {
            disposable.add(Completable.fromAction { binding.recyclerView.smoothScrollToPosition(0) }
                    .delay(200, TimeUnit.MILLISECONDS)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe { layoutManager.scrollToPositionWithOffset(0, 0) })
        } else {
            super.onBackPressed()
            finishAffinity()
        }
    }

    override fun onDestroy() {
        disposable.clear()
        super.onDestroy()
    }
}
